src/callouts.rs

```
@@ -0,0 +1,354 @@
+            }
+        }
+        if stdout {
+            let st = String::from_utf8_lossy(&output.stdout);
+            if !st.is_empty() {
+                let s = format!("{}: {}", &sname, st);
+
+                if !self.use_syslog {
+                    print!("{}", &s);
+                } else {
+                    let _ = logging::journal_print(Priority::Notice, &s);
+                }
+            }
+        }
+    }
+
+    fn invoke_first_matching_script<P: AsRef<Path>>(
+        &self,
+        dev: &mut MDev,
+        dir: P,
+        event: Event,
+        action: Action,
+    ) -> Option<(PathBuf, Output)> {
+        if dir.as_ref().read_dir().ok()?.count() == 0 {
+            return None;
+        }
+
+        for s in dir.as_ref().read_dir().ok()? {
+            let path = s.ok()?.path();
+
+            match self.invoke_script(dev, &path, event, action).ok() {
+                Some(res) => {
+                    if res.status.code().is_none() {
+                        warn!("Script was terminated by a signal");
+                        continue;
+                    } else if res.status.code() != Some(2) {
+                        return Some((path, res));
+                    } else {
+                        debug!(
+                            "Device type {} unmatched by callout script",
+                            dev.mdev_type().ok()?
+                        );
+                    }
+                }
+                _ => {
+                    debug!("failed to execute callout script {:?}", path);
+                    continue;
+                }
+            }
+        }
+        None
+    }
+
+    fn callout(&mut self, dev: &mut MDev, event: Event, action: Action) -> Result<()> {
+        let dir = dev.env.callout_script_base();
+
+        if !dir.is_dir() {
+            return Ok(());
+        }
+
+        let rc;
+
+        if let Some(ref s) = self.script {
+            let output = self.invoke_script(dev, s, event, action)?;
```

Here, I think we want to avoid using the `?` operator and returning an error from `callout()` if `invoke_script()` fails. We only want to return an error (and block primary command execution) if the script successfully executes and returns a valid response code. This prevents buggy scripts from disrupting mdevctl's functionality.